add ResponseErrors selector to router response

Aguilarjaf opened the pull request:
<!-- start metadata -->
Introducing the `ResponseErrors` selector in telemetry configurations to capture router response errors, allowing users to capture and log errors encountered at the router service layer. This new selector enhances logging for the router service, as it allows users the option to only log router errors instead of the entire router response body to reduce noise.

``` yaml
telemetry:
  instrumentation:
    events:
      router:
         router.error:
            attributes:
               "my_attribute":
                   response_errors: "$.[0]"
                   # Examples: "$.[0].message", "$.[0].locations", "$.[0].extensions", etc.
```

<!-- [ROUTER-7882] -->
---
**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] PR description explains the motivation for the change and relevant context for reviewing
- [ ] PR description links appropriate GitHub/Jira tickets (creating when necessary)
- [ ] Changeset is included for user-facing changes
- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- [ ] Metrics and logs are added[^3] and documented
- Tests added and passing[^4]
    - [x] Unit tests
    - [ ] Integration tests
    - [ ] Manual tests, as necessary

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: A lot of (if not most) features benefit from built-in observability and `debug`-level logs. Please read [this guidance](https://github.com/apollographql/router/blob/dev/dev-docs/metrics.md#adding-new-metrics) on metrics best-practices.
[^4]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

apollo-cla commented:
@Aguilarjaf: Thank you for submitting a pull request! Before we can merge it, you'll need to sign the Apollo Contributor License Agreement here: https://contribute.apollographql.com/

Velfi commented on apollo-router/src/plugins/telemetry/config_new/router/selectors.rs:
```diff
@@ -216,7 +226,7 @@ impl Selector for RouterSelector {
             RouterSelector::Static(val) => Some(val.clone().into()),
             RouterSelector::StaticField { r#static } => Some(r#static.clone().into()),
-            RouterSelector::ResponseBody { response_body } if *response_body => {
+            RouterSelector::ResponseBody { response_body, truncate: _truncate } if *response_body => {
```
Since we don't use `truncate` here, this should be fine:
```suggestion
            RouterSelector::ResponseBody { response_body, .. } if *response_body => {
```

Aguilarjaf commented:
Hi @goto-bus-stop, thanks for the feedback! 

I have updated the PR to better align with the best practice of using JSON Path to select a subset of data from the router response. I have implemented an additional selector, `ResponseErrors`, that takes a `JsonPathInst` object to extract specific information from router response errors (message, extensions, etc.). This is functionality currently does not exist in the router service response, so providing a new selector such as `ResponseErrors` can help users better capture and log those router level errors.

Velfi commented:
@mergifyio copy dev

abernix commented:
We'll close this in lieu of #8024, where you're welcome to help contribute to the conversation in getting it over the line but will no longer have branch push perms.  Thanks much for getting this opened and started originally — you'll still get contribution credit via that other PR when it merges, but we'll be able to run this with the full CI from that PR via @Velfi's account.  Thanks again to you both!
